happi/cli.py

```diff
@@ -392,6 +426,389 @@ def transfer(ctx, name: str, target: str):
+        context_profiler = None
+    elif profiler == 'cprofile':
+        context_profiler = Profile()
+
+        @contextmanager
+        def profiler_context(*args, **kwargs):
+            context_profiler.enable()
+            yield
+            context_profiler.disable()
+
+    @contextmanager
+    def null_context(*args, **kwargs):
+        yield
+
+    def output_profile():
+        # Call at the end: let's output results to stdout
+        if profiler == 'pcdsutils':
+            from pcdsutils.profile import print_results
+            print_results()
+        elif profiler == 'cprofile':
+            context_profiler.print_stats(sort='cumulative')
+        print('Profile completed successfully')
+
+    # Profile stage 1: searching the happi database
+    logger.info('Searching the happi database')
+    if profile_database:
+        db_context = profiler_context
+    else:
+        db_context = null_context
+    start = time.monotonic()
+    with db_context(
+        module_names=['happi'],
+        use_global_profiler=True,
+        output_now=False,
+    ):
+        if search_criteria:
+            items = search_parser(
+                client=client,
+                use_glob=use_glob,
+                search_criteria=search_criteria,
+            )
+        else:
+            # All the items
+            items = client.search()
+    logger.info(
+        f'Searched the happi database in {time.monotonic() - start} s'
+    )
+
+    if not any((profile_import, profile_object)):
+        return output_profile()
+
+    # Profile stage 2: import the device classes
+    logger.info('Importing the device classes')
+    if profile_import:
+        import_context = profiler_context
+    else:
+        import_context = null_context
+
+    classes = set()
+    start = time.monotonic()
+    with import_context(
+        module_names=['happi'],
+        use_global_profiler=True,
+        output_now=False,
+    ):
+        for search_result in items:
+            try:
+                cls = happi.loader.import_class(
+                    search_result.item.device_class,
+                )
+                classes.add(cls)
+            except ImportError:
+                logger.warning(
+                    f'Failed to import {search_result.item.device_class}'
+                )
+    logger.info(
+        f'Imported the device classes in {time.monotonic() - start} s'
+    )
+
+    if not profile_object:
+        return output_profile()
+
+    # Check which modules to focus on for line profiler
+    module_names = set(('happi',))
+    for instance_class in classes:
+        try:
+            parents = instance_class.mro()
+        except AttributeError:
+            # E.g. we have a function
+            parents = [instance_class]
+        for parent_class in parents:
+            module = parent_class.__module__
+            module_names.add(module.split('.')[0])
+    # All all ophyd control layers
+    module_names.add('epics')
```

Comment:
This is pretty specific to our use of happi - do these fail gracefully if the import isn't found?

Reply:
Yes, the bits in pcdsutils are very permissive with failures and just move on. https://github.com/pcdshub/pcdsutils/blob/69cf8f7ad027e13ef56c5ebe0098aec4b0e93597/pcdsutils/profile.py#L456-L476

Reply:
Is there a more natural way to figure out what modules to add? Adding everything seems to have consequences.

Reply:
Easiest would be to just leave it up to the user, though this is convenient for us for sure.

Random idea:
```python
happi_modules = set(sys.modules)
import_device()
modules_to_profile = set(sys.modules) - happi_modules
```

Reply:
Oh well, manual it is, I suppose?